### ISSUE TITLE
Hot Code Push broken on iOS

### ISSUE THREAD

**Issue by EliArtist** [meteor/meteor#10277]
My iOS app can't use hot code push to get new versions I upload. After some searching I believe I can narrow down the problem to a bug in cordova-plugin-meteor-webapp which tries to link the same file twice. I'm pretty sure that is what breaks hot code push.
I found a similar issue here: https://github.com/meteor/cordova-plugin-meteor-webapp
But that repository is untouched since February so I decided to open a new issue here.

My current Meteor version is 1.7.1-rc.5, I'll try to update the meteor version to 1.8, test again and post an update here.

EDIT:
After trying to build an iOS App with Meteor 1.8 I'm getting the same error

**Comment by dovydaskukalis** [comment]
Same issue, it stops working every time we update font files. Same error message as in https://github.com/meteor/cordova-plugin-meteor-webapp/issues/56.

**Comment by menelike** [comment]
Also related to https://github.com/meteor/cordova-plugin-meteor-webapp/pull/59

Right now we need to release a new app version on font updates.

**Comment by fizx1** [comment]
I was about to post a separate bug report, but ours seems like it could be the same issue. I'll post our details below. We are eager to see this working and willing to help out however possible. Thanks!

We are having some issues with Hot Code Push. In short, we are deploying to Galaxy using a Windows PC. We use that same PC to build the Android/Cordova apk. Hot Code Push works for these Android clients. Hot Code Push doesn't work, though, for iOS clients -- the iOS client is built on another machine. We are using the METEOR_CORDOVA_COMPAT_VERSION system, so it should work, from my understanding. There are no apparent error messages (maybe I'm not looking in the right place), but when I redeploy a new app version to Galaxy, only Android clients process the update. Thanks in advance and please let me know how I can help further.
The long version...
We are deploying from a Windows machine to Galaxy using the following commands:
```
SET DEPLOY_HOSTNAME=galaxy.meteor.com
SET METEOR_CORDOVA_COMPAT_VERSION_ANDROID=v0001
SET METEOR_CORDOVA_COMPAT_VERSION_IOS=v0001
meteor deploy appname.meteorapp.com --settings settings.json --mobile-server=https://appname.meteorapp.com
```

The hot code push version number is confirmed to be making it to the Galaxy server; we can see the file http://appname.meteorapp.com/__cordova/manifest.json is visible and contains the entry:
"cordovaCompatibilityVersions":{"android":"v0001","ios":"v0001"}

We are building for iOS App Store with the following command:
`METEOR_CORDOVA_COMPAT_VERSION_IOS=v0001 meteor build build_directory --server https://appname.meteorapp.com`

After building, we open the build in XCode, create an archive, and upload the archive to the App Store for approval and release. 
(please PM me for the actual appname, thanks!)

**Comment by rj-david** [comment]
Anybody figured how to workaround this bug?

**Comment by sugoke** [comment]
Same issue here. I have an app deployed on heroku, and when I build for iOS, HCP seems to be broken. I have seen a few messages regarding this “bug” on other forums but no real explanation.

`**2018-10-18 12:40:36.942203+0200 Premia[1933:140955] Serving asset bundle version: 3e53d8a74f99d185edd94272536b033999610f80**

**2018-10-18 12:40:39.200604+0200 Premia[1933:140955] Start downloading asset manifest from: manifest.json -- https://premia.herokuapp.com/__cordova/**

**2018-10-18 12:40:40.369383+0200 Premia[1933:141234] Downloaded asset manifest for version: 9c4c53cdc986cd3eb49b9f99b097a1bb76305b42**

**2018-10-18 12:40:40.370495+0200 Premia[1933:141234] Download failure: Skipping downloading blacklisted version**

**2018-10-18 12:40:40.374600+0200 Premia[1933:140955] ERROR: {"line":36,"column":30,"sourceURL":"http://localhost:12224/plugins/cordova-plugin-meteor-webapp/www/webapp_local_server.js"}**`

**Comment by rj-david** [comment]
@benjamn, we would like to ask some guidance regarding this?

@RealHandy seems to have some clues on what's happening here: https://github.com/meteor/cordova-plugin-meteor-webapp/issues/56

**Comment by rj-david** [comment]
Just saw these related PRs - hope they made it to 1.8.1.beta soon

https://github.com/meteor/cordova-plugin-meteor-webapp/pull/61
https://github.com/meteor/cordova-plugin-meteor-webapp/pull/62
https://github.com/meteor/meteor/pull/10219

**Comment by Wade-BuildOtto** [comment]
I am seeing the same issue on 1.7.0.4 with METEOR_CORDOVA_COMPAT_VERSION_IOS in use, its ignoring it even though what is listed on the manifest is accurate. I might role back to see if that helps, but it's starting to get costly rolling out small fixes via app store updates.

**Comment by rj-david** [comment]
The PRs related to the cordova-plugin-meteor-webapp has been merged to master! Wohoo!

For those who wanted to follow the PR for the main meteor project related to HCP and other cordova stuff, subscribe to this PR:
https://github.com/meteor/meteor/pull/10339

**Comment by RealHandy** [comment]
This is also #10181

**Comment by Wade-BuildOtto** [comment]
has anyone tried meteor update --release 1.8.1 to see if the update fixes the issue. 

"Not it."

**Comment by Wade-BuildOtto** [comment]
any movement on this?

**Comment by sugoke** [comment]
Interestingly, the problem occurs only when deployed (to Heroku) on my side. On localhost, HCP works. 1.8.0.1 did not fix.

**Comment by macrozone** [comment]
> Interestingly, the problem occurs only when deployed (to Heroku) on my side. On localhost, HCP works. 1.8.0.1 did not fix.

yes, on develop it works, but not once you deployed it.

**Comment by macrozone** [comment]
> I upgraded from 1.6 to 1.8, and HCP didn't work even though I used the compatibility override. I had to immediately rollback.
> 
> Is this the correct issue for what I experienced?

yes, i think so. I noticed too late and can't roll back easily now :-(

**Comment by bmanturner** [comment]
That's unfortunate... I had a huge mongo migration (servers and schemas) as a part of this deployment, so luckily I was able to revert the merge and go back to using the old mongo instance.

**Comment by sugoke** [comment]
Did Meteor team comment on that?

It is quite an important feature of the framework.


Le 19 déc. 2018 à 16:45, Brendan Turner <notifications@github.com<mailto:notifications@github.com>> a écrit :

That's unfortunate... I had a huge mongo migration (servers and schemas) as a part of this deployment, so luckily I was able to revert the merge and go back to using the old mongo instance.

**Comment by RealHandy** [comment]
Yes, this is all the same issue: see also #10181. The underlying issue is at https://github.com/meteor/cordova-plugin-meteor-webapp/issues/56

**Note: I've just uploaded what I think is an improved workaround:** meteor/cordova-plugin-meteor-webapp#59

If you know how to get the forked PR version, put it on your local drive, and make that the version in your project, you can use this to fix the problem, but it does require that you push a new iOS version out to users with the fix in it.

**Comment by RealHandy** [comment]
@bmanturner This bug is related to changes to assets. If the upgrade causes some asset change, then the bug will bite you on iOS and hot code push will fail (as described in the posts across the various related issues and as seen in @macrozone's post above (https://github.com/meteor/meteor/issues/10277#issuecomment-446917778). If HCP is failing, the workaround can only be applied by an update to the user's phone version, since the bug fix is in cordova code.

**Comment by rj-david** [comment]
Fix to iOS HCP has been committed. Excited to test this early next week

https://github.com/meteor/meteor/pull/10248#issuecomment-453613510

**Comment by benjamn** [comment]
Thanks especially to @RealHandy for finding an elegant solution!

**Comment by rj-david** [comment]
Thanks @benjamn and @RealHandy. Great way to start 2019. Hope for the stable release of 1.8.1 soon

**Comment by rj-david** [comment]
@benjamn, I made a test on a production app:
1. Updated meteor version to 1.8.1-beta.13
2. Build and upload to apple test flight, and deploy to our server i.e.. version 1.5.0
3. iOS app successfully updated
4. Made changes to the build, tag a new version and deploy to server i.e. version 1.5.1

The iOS app is still not picking up the new build. I also tested a forced refresh `window.location.reload(true)` but it did not pick up the new build.

We are preparing to test now with Android (it was not an issue with Android previously but testing it anyway)

**Comment by rj-david** [comment]
Ok, Android hot code push is also not working on our builds. We will now be doing a comparison between 1.8.1-beta.13 and 1.8.1-beta.12

**Comment by rj-david** [comment]
I deleted my previous replies due to new data we collected

Confirming that Hot Code Push now works with iOS with one caveat. There is a need for a complete uninstall of any existing app. When we tried updating an existing iOS app, HCP did not work even with the fix. When we tried to uninstall the app and do a fresh install of the iOS with the app, HCP worked.

(...testing android)

**Comment by rj-david** [comment]
HCP is not working in Android with Meteor 1.8.1-beta.13 (tested on multiple android phones)

We are using the `Reload.isWaitingForResume()` feature and it can detect that a new version is available. But once we did a refresh, the new version is not being successfully loaded.

**Comment by menelike** [comment]
@rj-david 

If I understand https://github.com/meteor/cordova-plugin-meteor-webapp/pull/62 correctly, https://github.com/meteor/meteor/pull/10219 needs to merged to stay backward compatible.

Do you have any logs, hard to help without them.

**Comment by rj-david** [comment]
I can confirm now that HCP works for both android and iOS using Meteor 1.8.1-beta.14

**Comment by RealHandy** [comment]
@rj-david That’s great!

**Comment by rj-david** [comment]
Is this related to any of the changes for beta.14?
https://github.com/meteor/meteor/issues/10426

**Comment by RealHandy** [comment]
The HCP fix itself wouldn’t be related, it doesn’t do anything with those configs.

**Comment by rj-david** [comment]
@RealHandy, @benjamn 

We are actively developing an app in beta and launched last Monday using Meteor 1.8.1-beta.17. We found an issue with HCP in iOS with the latest fix. The first HCP in iOS worked without problem every time. After that the HCP stopped working again.

**Comment by menelike** [comment]
@rj-david Are you able to provide some logs from the affected devices?

**Comment by rj-david** [comment]
> @rj-david Are you able to provide some logs from the affected devices?

We don't have any. What is the best way to get these logs?

**Comment by RealHandy** [comment]
@rj-david You connect your iPhone to your Mac with a usb, then you run Xcode and from the top menu select Window - > Devices , then select your connected device from left side list. The current log will show scrolling at the bottom of the screen, though you may need to hit the little arrow at the bottom to expand logs. Then you reproduce the problem and it will show in the logs. A lot of other apps and system stuff may be filling the logs, so you’re likely to need to filter the log by something specific to your app (or just have good eyes).

**Comment by rj-david** [comment]
@RealHandy, I'm not 100% sure if we can replicate this for a development setup. Our team has been developing with an iPhone connected for quite some time and nobody raised an issue (I was using Ubuntu + Android though). We just noticed the problem when we started releasing in production. In all versions, the iOS app will HCP once when a new version is deployed. For the next version, it won't.

I'll check with those in our team using iphone and mac tomorrow.

Anybody knows if APM can catch this and what should we look at in the apm logs?

**Comment by rj-david** [comment]
This is the only relevant logs that we can see as related (i.e. the BLACKLIST part)

```
2019-02-18 12:54:32.961450+0800 Bountee[15907:7331347] App startup confirmed
2019-02-18 12:54:41.794419+0800 Bountee[15907:7331347] Start downloading asset manifest from: manifest.json -- http://192.168.1.18:3003/__cordova/
2019-02-18 12:54:41.848020+0800 Bountee[15907:7332764] Downloaded asset manifest for version: 38a4c717e7e8c577e9eac4a97566cd3c0013001b
2019-02-18 12:54:41.848491+0800 Bountee[15907:7332764] BLACKLIST - blacklistedVersions: []
2019-02-18 12:54:41.848580+0800 Bountee[15907:7332764] Finished downloading new asset bundle version: 38a4c717e7e8c577e9eac4a97566cd3c0013001b
2019-02-18 12:54:41.883098+0800 Bountee[15907:7331347] Serving asset bundle version: 38a4c717e7e8c577e9eac4a97566cd3c0013001b
2019-02-18 12:54:42.886473+0800 Bountee[15907:7331347] App startup confirmed
2019-02-18 12:54:42.921297+0800 Bountee[15907:7331347] Start downloading asset manifest from: manifest.json -- http://192.168.1.18:3003/__cordova/
2019-02-18 12:54:42.979222+0800 Bountee[15907:7332978] Downloaded asset manifest for version: 38a4c717e7e8c577e9eac4a97566cd3c0013001b
```

**Comment by RealHandy** [comment]
This shows you’re successfully getting the asset manifest and the app is resetting due to having gotten it. Then it gets the asset manifest in the restarted version. All of that looks as it should be and there are no errors due to asset update. I don’t know what the blacklist means, it has no versions in its list, and the app is proceeding to restart and function after that. There’s no error here that I can see, unless it’s related in some way to the blacklist. This log has no issue related to the original #10277 bug. Bummer, not much to go on here.

**Comment by KoenLav** [comment]
@rj-david you mentioned in the 1.8.1 release thread that the second HCP on iOS fails consistently.

Do others experience the same thing?

@wojtkowiak maybe you could chime in?

**Comment by rj-david** [comment]
@KoenLav, yes, we were able to confirm it multiple times in our project.

Just a few minutes ago, we released a new version without the mdg:reload-on-resume package to test if that was the one causing the issue. Unfortunately with iOS, we have an earlier release with a change in mobile-config.js which meant we have to go through the App Store review. We expect it to be approved within 12 hours from this post. Will post here if we find something useful.

**Comment by mozfet** [comment]
> @rj-david you mentioned in the 1.8.1 release thread that the second HCP on iOS fails consistently.
> 
> Do others experience the same thing?
> 
> @wojtkowiak maybe you could chime in?

I am definitely experiencing it. I've been unable to deploy iOS for more than a month. Works fine locally, but not in App Store. Could not find good logs.

**Comment by rj-david** [comment]
> I am definitely experiencing it. I've been unable to deploy iOS for more than a month. Works fine locally, but not in App Store. Could not find good logs.

@mozfet, are you using at least 1.8.1-beta.14 ?

**Comment by mozfet** [comment]
> > I am definitely experiencing it. I've been unable to deploy iOS for more than a month. Works fine locally, but not in App Store. Could not find good logs.
> 
> @mozfet, are you using at least 1.8.1-beta.14 ?

Yup - METEOR@1.8.1-beta.14

**Comment by buymybm100** [comment]
I am using METEOR@1.8.1-beta.14 and been pushing new builds to my HockeyApp Test environment. The hot push fails the first time and all subsequent times - on IOS and Android.

**Comment by rj-david** [comment]
@buymybm100, how did you deploy your ipa and apk builds? Testflight? Beta channels in play store console?

**Comment by RealHandy** [comment]
I’m on my fork of the plugin and 1.8, and HCP works fine for any number of HCP updates. When I get a chance, I’ll debug 1.8.1 and see if I can see anything. It’s worth mentioning that I had to patch a 3rd party plugin because the “restart” of HCP caused a crashing bug, so other plugins could be an issue for HCP.

**Comment by rj-david** [comment]
@RealHandy, what 3rd party plugin did you need to patch? Might give a clue on this

**Comment by RealHandy** [comment]
Heh, indooratlas — def not something anyone else is using.

**Comment by buymybm100** [comment]
> @buymybm100, how did you deploy your ipa and apk builds? Testflight? Beta channels in play store console?

Through Hockeyapp as it can host apps for both IOS and Android.

**Comment by brucejo75** [comment]
Hi All,

Are you connecting to your devices and examining the console output?  There are a number of legitimate reasons why HCP will fail:
- if there is an issue on startup that causes any `Meteor.startup` handler to fail or run too slowly then the build will be marked as an error build and Meteor won't try to download it again.  The console will show this case.
- Meteor detects that the cordova packages are different for the HCP than the installed app and refuses to load the HCP.  The console will show this case.
- Builds of `cordova-plugin-meteor-webapp prior` to 1.6.5 had an [issue with partial downloads](https://github.com/meteor/meteor/issues/9387)  which was fixed with [PR 38](https://github.com/meteor/cordova-plugin-meteor-webapp/pull/38) which is the default for `v1.8.1`.  The console shows the app has failed at a really weird place, upon inspection you can see that `app.js` is incomplete.

**Comment by rj-david** [comment]
@Nauzer, I am replying here instead from the release PR thread.

Reference from other folks here. @Nauzer did a test showing that there was nothing wrong with the HCP in production. Here is the link: https://github.com/meteor/meteor/pull/10248#issuecomment-469680616

For the question regarding my plugins, here they are:
```
branch-cordova-sdk@3.1.5
cordova-plugin-camera@4.0.3
cordova-plugin-datepicker@0.9.3
cordova-plugin-device@2.0.2
cordova-plugin-file@6.0.1
cordova-plugin-filepath@1.5.1
cordova-plugin-geolocation@4.0.1
cordova-plugin-inappbrowser@3.0.0
cordova-plugin-ionic-keyboard@2.1.3
cordova-plugin-nativestorage@2.3.2
cordova-plugin-splashscreen@5.0.2
cordova-plugin-statusbar@2.4.2
cordova-plugin-vibration@3.1.0
cordova-plugin-wkwebview-engine@1.1.4
cordova-plugin-x-socialsharing@5.4.4
onesignal-cordova-plugin@2.4.6
phonegap-plugin-barcodescanner@8.0.1
```

**Comment by rj-david** [comment]
> Are you connecting to your devices and examining the console output?

@brucejo75, we cannot replicate it in development. Happening for us in production. What is the best way to get any logs out from production?

**Comment by mozfet** [comment]
> > @buymybm100, how did you deploy your ipa and apk builds? Testflight? Beta channels in play store console?
> 
> Through Hockeyapp as it can host apps for both IOS and Android.

It breaks for me when loading it with Testflight using my STAGE environment. It does not happen in Dev, and I can use the Browser as normally; so there should be no build or dependency issues.

**Comment by mozfet** [comment]
I can also load the App directly onto my Dev Android and iOS devices and it works fine, so I do not expect any Cordova issues either... not that I can see any of this in PROD, since debug logs are not available and it does not proceed past the Splash Screen.

Is it possible to connect a debugger to an iOS App deployed using Testflight?

**Comment by menelike** [comment]
> Happening for us in production. What is the best way to get any logs out from production?

@rj-david You can build your project as a developer build (instead of distribution for app store), you'll be able to install that on your device directly and see the log files.

**Comment by mozfet** [comment]
It does not happen in DEV when you load it directly on the device using XCode, so there is nothing to log.

**Comment by rj-david** [comment]
> @rj-david @mozfet You can build your project as a developer build (instead of distribution for app store), you'll be able to install that on your device directly and see the console log.

We are doing this now. I'll post here the results once we are done

**Comment by rj-david** [comment]
@wildhart was able to replicate our experience: https://github.com/meteor/meteor/pull/10248#issuecomment-470054560

The plugins that were the same for our builds are as follows:
cordova-plugin-device@2.0.2
cordova-plugin-file@6.0.1
cordova-plugin-inappbrowser@3.0.0
cordova-plugin-splashscreen@5.0.2
cordova-plugin-statusbar@2.4.2

**Comment by menelike** [comment]
@mozfet 

What exactly do you mean with `DEV`? The local meteor development (aka http://localhost:3000) or a dev bundle targeting your production instance?

**Comment by Nauzer** [comment]
Hi @rj-david @mozfet @wildhart

Try adding the latest version of cordova-plugin-meteor-webapp to your cordova plugins list. 
`cordova-plugin-meteor-webapp@1.6.5`

I think underwater Meteor is still using 1.6.0 and I manually set it to a higher version in my Cordova plugins file. If that is the case we might need to do a PR for 1.8.1 release to update cordova-plugin-meteor-webapp to it's latest version.

**Comment by Nauzer** [comment]
Looks like @wildhart has already done this... 
As I can confirm I do **not** experience this issue I post hereby my cordova-plugins & package versions:

https://pastebin.com/Lz3MnnpB

**Comment by mozfet** [comment]
@menelike 

- ```DEV``` as in development environment using local machines and Xcode and running in Similators and Sideloading on Devices.
- ```STAGE``` aka ```TEST``` and ```PRE_PROD``` environment, where we use test channels representing the real world, and a temp domain name.
- ```PROD``` as in production environment, using official .com domain, App Store and Play Store Distribution.

**Comment by menelike** [comment]
@mozfet This is what I was trying to explain, you can run a dev build (development in terms of Xcode) against your production. The app should logically behave the same, except of debugging features. 

When we want to understand/debug what is happening on production, we use our app dev build against our production servers.

**Comment by mozfet** [comment]
@menelike I have run dev build and sideloaded it to talk with our ```STAGE``` server, which is the same as ```PROD``` server will become (but on a different domain). It works fine, no issues. Only once it is loaded using Test Flight to also talking with our ```STAGE``` server, I have app level logs, but they are not shown and the app does not progress beyond the splash screen - thus no logs...

**Comment by rj-david** [comment]
Ok, we got something. Hopefully this is related

```
2019-03-06 20:46:08.897897+0800 Bountee[1257:324345] ERROR: checkForUpdates requires a rootURL to be configured
2019-03-06 20:46:10.660483+0800 Bountee[1257:324345] App startup confirmed
2019-03-06 20:46:10.747099+0800 Bountee[1257:324345] ERROR: checkForUpdates requires a rootURL to be configured
```

**Comment by menelike** [comment]
@rj-david This seems to be a configuration issue. Have you built your project from meteor just like you would for production?

**Comment by Nauzer** [comment]
in your Settings file do you have a ROOT_URL key?

{
    "public": { ... },
    "private": {
        "ROOT_URL": "https://YOURDOMAIN"
    }
}

**Comment by rj-david** [comment]
> @rj-david This seems to be a configuration issue. Have you built your project from meteor just like you would for production?

Yes. The app was running ok and HCP'd initially for the first server update. On the 2nd server update, the errors starting to occur right after the new build became live in production server

**Comment by rj-david** [comment]
> @rj-david This seems to be a configuration issue. Have you built your project from meteor just like you would for production?

Yes. The first deployment to production server, the HCP worked. The next deployment to production server, that's when HCP stopped working and we got the error.

**Comment by menelike** [comment]
> https://app.bountee.com/__cordova/__cordova/082522452f0ed7b03ab15cb4f71b3f33369e7c6e.stats.json?meteor_dont_serve_index=true

Just by looking at this, why does it include `__cordova/__cordova` (2x times)?
We need to check if the URLs are generated correctly.

**Comment by rj-david** [comment]
> in your Settings file do you have a ROOT_URL key?
> Can you share your Settings file?
> Did you change anything in that between the two updates?
> 
> {
> "public": { ... },
> "private": {
> "ROOT_URL": "https://YOURDOMAIN"
> }
> }

We don't. We have the ROOT_URL in the server i.e. bountee.service file

Could this be the cause of our issue?

**Comment by rj-david** [comment]
> I have no idea why this is the result, usually a json should be responded.
> 
> https://app.bountee.com/__cordova/manifest.json outputs `https://app.bountee.com/__cordova/9a5292ccde3010222a53a8be41aeb0ad65a60e1f.stats.json?meteor_js_resource=true` which doesn't respond with a JSON as well.

Hope this can lead us somewhere.

(My team just went home - past 9pm now locally; if there will be anything required from us, we will work on it tomorrow as soon as we are back in the office)

**Comment by menelike** [comment]
@rj-david 

How do you build your app? We use `meteor build outputDir --server=https://prod.server --mobile-settings settings.json`. AFAIK ROOT_URL should not be affecting production app builds at all (of course), hence I don't understand your logs. Maybe a meteor reset might be run before you build, I've seen traces of development stuff when bundling for production in the past.

**Comment by rj-david** [comment]
@menelike

We build similar to what you posted. Normally, it is being built using our CI box but for this test, I built it manually. Tomorrow, I'll delete the local folder and build again

Any idea on what to check on our server setup? Note that we don't have a problem with Android.

The settings values are all being used in the app. The server settings are being downloaded separately from a different location which is automatically downloaded by our ASG in AWS when scaling or deploying (as part of our CI/CD pipeline)

**Comment by menelike** [comment]
@rj-david 
You've got a good point with: _Note that we don't have a problem with Android._

As this point this has become a guessing game, still I'd try to understand the URLs first, even if this is not the causing the issue (`__cordova/__cordova` is just too suspicious).

**Comment by wildhart** [comment]
I'm building/deploying with MUP.  HCP works for me every time with Android, but on iOS only works the first time after a fresh app install from the App Store.  I can't try loading a DEV build onto a device because I don't currently have access to a mac, so I can't get the logs, sorry.

ROOT_URL is set within MUP: `ROOT_URL: 'https://www.virtualinout.com'`

Here's my manifest if that helps: https://www.virtualinout.com/__cordova/manifest.json

I'm already using cordova-plugin-meteor-webapp@1.6.5: https://pastebin.com/yYvNG9sK

**Comment by rj-david** [comment]
@RealHandy, you might be more familiar with this:

https://github.com/meteor/meteor/issues/10277#issuecomment-470095469

**Comment by rj-david** [comment]
I'm on a mobile phone and difficult to check further but here is the start:

```
private void checkForUpdates(final CallbackContext callbackContext) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                HttpUrl rootUrl = HttpUrl.parse(currentAssetBundle.getRootUrlString());
                if (rootUrl == null) {
                    callbackContext.error("checkForUpdates requires a rootURL to be configured");
                    return;
                }
                HttpUrl baseUrl = rootUrl.resolve("__cordova/");
                assetBundleManager.checkForUpdates(baseUrl);
                callbackContext.success();
            }
        });
    }
```

https://github.com/meteor/cordova-plugin-meteor-webapp/blob/master/src/android/WebAppLocalServer.java

**Comment by rj-david** [comment]
How/Where do you guys set your ROOT_URL?

@Nauzer?

**Comment by rj-david** [comment]
> > https://app.bountee.com/__cordova/__cordova/082522452f0ed7b03ab15cb4f71b3f33369e7c6e.stats.json?meteor_dont_serve_index=true
> 
> Just by looking at this, why does it include `__cordova/__cordova` (2x times)?
> We need to check if the URLs are generated correctly.
> 
> ### Update
> I can confirm that the URL scheme `https://foo.bar/__cordova/HASH.stats.json` should resolve.

The URL of the file should have been:
https://app.bountee.com/__cordova/770229e9f890607115d0f9295b325e97c3bf52d5.stats.json?meteor_js_resource=true

So (1) wrong url path (2x cordova folder), and (2) wrong hash

(Note that the hash might change moving forward as the team will start deploying as we work today. But at the time of posting, the URL I posted works)

**Comment by rj-david** [comment]
Made a 2nd test today for iOS HCP. Once again, the first HCP worked, the 2nd HCP did not. But this time, there aren't any error messages. The only relevant log messages were
```
2019-03-07 14:53:43.255402+0800 Bountee[1364:373200] Start downloading asset manifest from: manifest.json -- https://app.bountee.com/__cordova/
2019-03-07 14:53:43.262244+0800 Bountee[1364:376514] Downloaded asset manifest for version: 95d7e443cacd42bd85f32552c44af2f88520bdae
```

**Comment by mozfet** [comment]
@rj-david I use a settings-production.json with Galaxy settings to deploy to Galaxy and do not set a ROOT_URL... should I?

**Comment by Nauzer** [comment]
@mozfet Could you try that?
I do have this setting - make sure you include the setting (private.ROOT_URL) in your initial app build and on your galaxy server build using the following steps.
I always add the first 3 lines as I've seen old settings files "sticking around" when not doing this.

```
rm -rf .meteor/local/cordova-build
meteor remove-platform ios
meteor add-platform ios
meteor build --mobile-settings=yoursettingspath.js // mind the =
meteor deploy --settings yoursettingspath.js // no =
```

**Comment by mozfet** [comment]
@Nauzer 

Ok, do you mean I should make my ```settings-productions.json``` file should look like this?
```
{
  "galaxy.meteor.com": {
    "env": {
      "MONGO_URL": "...",
      "MONGO_OPLOG_URL": "...",
      "MAIL_URL": "...",
      "ROOT_URL": "https://www.mydomain.com",
    }
  }
}
```

I have an existing ```mobile-config.js``` in the root of my project folder, which used to work fine as default I do not expect I need to change that, and I then usually build with...
```
$ meteor build <BUILD_DIR> --server <SERVER_URL>
```

**Comment by Nauzer** [comment]
Nope - It should be part of your 'private' element I think.

{
  "public": {
    "param1": "abc"
  },
  "private": {
    "ROOT_URL": "https://yourdomain or local IP"
  },
  "galaxy.meteor.com": {
    "env": {
       "MONGO_URL": "...",
       "MONGO_OPLOG_URL": "...",
     }
  }
}

Please note that mobile-config.js is something different than the loaded settings file. Mobile config is picked up automatically. Settings file you need to explicitly add with above mentioned flags. 

Besides I build with:
meteor build "$1" --server=https://domain.com:443 --mobile-settings="$3"
Explicitly adding **:443** and with the = sign. Somehow that was proven to be necessary somewhere in the past can't remember why exactly but it might be of help now.

**Comment by RealHandy** [comment]
@rj-david Yep, WebAppLocalServer.swift is in the code I debugged to resolve the prior iOS HCP bug. The double __cordova in the failed path seems like a clear culprit in that log, although I almost feel like I’ve seen that before. I haven’t gone to 1.8.1 beta, so I don’t know whether I’d be able to repro this new issue or not.

**Comment by wildhart** [comment]
Just a thought, but could this line in [src/ios/AssetBundleManager.swift](https://github.com/meteor/cordova-plugin-meteor-webapp/blob/aa784445f2ed95a099af2d474386fe95f051da58/src/ios/AssetBundleManager.swift):
```js
211  // Workaround for https://github.com/meteor/cordova-plugin-meteor-webapp/issues/56
212 if assetBundle.assetExistsInBundle("/__cordova" + asset.urlPath) { continue }
```
Somehow be causing the downloading of the new bundle to be skipped?

**Comment by mozfet** [comment]
@Nauser Thank you. That seems to have fixed it for me. I will continue testing.

**Comment by RealHandy** [comment]
@wildhart that would only skip a new asset if that asset does NOT occur twice in manifest.json — once with the __cordova in the path and once without — and that would be a new manifest behavior. It could explain the double __cordova if asset.urlPath has a __cordova in it already, but I think that would mean the assets have an __cordova crested by the dev in the public assets dir. I should have time this week to look in on this issue.

**Comment by Nauzer** [comment]
@mozfet np. Looking forward to hear your results of HCP against your prod(-like) environment.

**Comment by rj-david** [comment]
@Nauzer, adding ROOT_URL in my private settings does not work in my case.

Regarding the manifest, I was able to do some further debugging now. I observed that the first time the HCP happens, seem like the iOS app caches manifest.json. The next time a new version of the app exists in the server, the version of the manifest being displayed in the console was still the old version.

First HCP console.log:
Start downloading assets from bundle with version: **ABC**

Second HCP console.log:
Start downloading assets from bundle with version: **ABC** <--- same version being displayed from the first HCP

If I uninstalled the app and reinstalled it, then the console will display a new version and the HCP will be successful:
Start downloading assets from bundle with version: **XYZ**

**Comment by rj-david** [comment]
Made two more tests and got the same result as above. The 2nd HCP seemed to be caching the manifest downloaded and the version did not change.

**Comment by Nauzer** [comment]
@rj-david sorry to hear. Just a try is that with a settings.json file you pass as a build parameter in both the Cordova as the deployed build (e.g. a file part of your bundle) or with this remote API file you referred to earlier?
If the latter, using a settings file as per the Meteor docs might be worth your while.

**Comment by brucejo75** [comment]
I am still working with `v1.7.0.5` of Meteor.  I have observed that the first download to a freshly installed Android App, does not complete.  And this is reproducible in my dev environment.

### Repro steps
In Development environment
1. device: Uninstall app.
2. device: Install app.
3. device: Start app.
4. Dev server: update the app for an HCP.

**Result:** HCP will happen (`Reload._onMigrate` fires), but the app running is not updated to the latest version.  It is still running the version that the app was built with.

### Observation
5. At this point, connect to the APP for debugging and from the console call `location.reload()`.

**Result:** the update will complete and my version is now the latest.

- Somehow the first HCP update is not completed properly.
- After the first HCP, subsequent HCPs work fine.
- I think this may have been happening for a while, but I just would bang on `location.reload()` as part of my dev process.

### Workaround
I have not tried (busy on other stuff) this but a way to encode this workaround would be:
```javascript
let needsToReload = localStorage.getItem('needsToReload');
Reload._onMigrate((retry) => {
  if(!needsToReload) localStorage.setItem(`needsToReload`, 'true');
});

if(needsToReload && needsToReload === 'true') {
  localStorage.setItem('needsToReload', 'false');
  location.reload();
}
```
This should only fire a `location.reload()` on the first HCP.
**Note:** this code needs to be on the installed APP.

**Comment by menelike** [comment]
@rj-david 
Some years back we had issues with HCP in conjunction with our web server (nginx). Do you have a web server in between? It *might* interfere with meteor when caching is enabled.

**Comment by rj-david** [comment]
@Nauzer, yes, through a settings.json file

@menelike, good point. We are using nginx. I'll ask our devops to exclude manifest.json from the cache and test again tomorrow.

**Comment by menelike** [comment]
@rj-david Just found the old thread: https://forums.meteor.com/t/app-constantly-refreshing-after-an-update/23586/113 might be outdated though.

**Comment by rj-david** [comment]
@brucejo75, before 1.8.1-beta, we did try variations of reloading/restarting the app but it only worked with Android. Beta.15 solved all issues we had with android so we removed the reload code since it wasn't working with iOS

**Comment by rj-david** [comment]
@menelike, I'm sure we have that cache since I was the one who initially set it. Will get back tomorrow for the results. Thanks

**Comment by wildhart** [comment]
We're also using nginx, deployed with MUP.  Haven't changed any ngnix settings from the MUP defaults. 
 Response headers on `https://www.virtualinout.com/__cordova/manifest.json` show `cache-control: public, max-age=31536000` but that should be served the same for Android and iOS.

**Comment by mozfet** [comment]
> @mozfet np. Looking forward to hear your results of HCP against your prod(-like) environment.

Looks like its working fine in my STAGE environment. Hold thumbs... going to try PROD as soon as Apple approved the app in the App Store.

**Comment by rj-david** [comment]
@menelike, removing the cache settings in our nginx solved this problem for us.

@Nauzer, @RealHandy, @benjamn, FYI.

Thanks everyone and cheers!

**Comment by menelike** [comment]
@rj-david puhh....finally, this was a hard one. Now everyone can be sure that HCP works again.

**Comment by wildhart** [comment]
So what would I need to change in my default MUP nginx config to fix my experience of this same issue?

MUP provides the `nginxServerConfig` setting but I wouldn't know what to change.  @zodern, have you heard of other MUP users having this issue (HCP on iOS only working the 1st time after a fresh app install. Android is fine. Previously experienced by @rj-david and myself above, but resolved for @rj-david by removing nginx config settings)?

**Comment by rj-david** [comment]
@wildhart, I took a quick look at MUP and seems like it has a default nginx setting. Since this is being used by a lot of meteor users, the understanding is that it has the correct nginx cache settings (or in this case, there are no cache settings).

Have you seen anything related to what we saw yesterday: https://github.com/meteor/meteor/issues/10277#issuecomment-471980580

**Comment by rj-david** [comment]
> We're also using nginx, deployed with MUP. Haven't changed any ngnix settings from the MUP defaults.
> Response headers on `https://www.virtualinout.com/__cordova/manifest.json` show `cache-control: public, max-age=31536000` but that should be served the same for Android and iOS.

Sorry, I missed this post. This cache setting was the one that was previously causing us problems

**Comment by wildhart** [comment]
I discovered the MUP command `mup proxy nginx-config` which gave this output: https://pastebin.com/bUcbC0Pz

I can't see anything there about cache except for SSL cache.

> Have you seen anything related to what we saw yesterday: [#10277 (comment)](https://github.com/meteor/meteor/issues/10277#issuecomment-471980580)

I haven't been able to access iOS app logs during an HCP yet, but I'll try.

**Comment by rj-david** [comment]
You might want to explicitly add a location block just for manifest with cache control header ignored: 
https://www.nginx.com/blog/nginx-caching-guide/

I'm not familiar how to use 'nginxServerConfig'

**Comment by Nauzer** [comment]
> ### Workaround
> I have not tried (busy on other stuff) this but a way to encode this workaround would be:
> 
> ```js
> codeblock
> ```
> 
> This should only fire a `location.reload()` on the first HCP.
> **Note:** this code needs to be on the installed APP.

@brucejo75 
Thanks for this tip. Works like a charm for us 👍 
Since I noticed that after the first reload the HTML/CSS is still the 'old' version of my app but the React components are actually already on the **new** version I thought I can make use of that 'feature' and implement as follows:

Created a Meteor method that checks a sent version against the current server version.
```js
reloadRequired: function(buildCode) {
        if (!buildCode || Meteor.settings.public.build > buildCode) {
            console.log('reload required for App with buildcode', buildCode);
            return true;
        } else {
            return false;
        }
    }
```

From a React component that is already loaded by the first version of my app on my login screen (the first screen of the app) I call this Method in the _componentDidMount()_ lifecycle function. 
```js
componentDidMount() {
        if(Meteor.isCordova && 'android' === device.platform.toLowerCase()) {
            Meteor.setTimeout(() => {
                Meteor.call('reloadRequired', Meteor.settings.public.build, (e,result) => {
                    if(!e && result) {
                        location.reload();
                    }
                });
            }, 10000);
        }
    }
```

Note that the build code already was part of my settings file used to build the initial App Store version of the app. But even if you don't have that implemented yet I believe you can use the above code as long as you have an **existing** React component - preferably on the first page of your app.

**Comment by rj-david** [comment]
@wildhart, were you able to solve the cache?

**Comment by wildhart** [comment]
No, I haven't been able to play with the nginx cache yet, and I'm on vacation now until April..

**Comment by wildhart** [comment]
Sorry for the haitus.  I've been able to access the iOS logs and confirmed the findings of @rj-david https://github.com/meteor/meteor/issues/10277#issuecomment-471980580, that iOS is caching the `/__cordova/manifest.json` file after the first HCP.

Using MUP I've managed to prevent the cache and finally got HCP working on iOS again.  See my MUP bug report/suggestion here: https://github.com/zodern/meteor-up/issues/1086

**Comment by KoenLav** [comment]
@wildhart we seem to have resolved our issue using a similar approach using Cloudflare (setting a Page rule which bypasses the cache for any requestions to *.ourdomain.com/*/manifest.json).

@benjamn I'm guessing it should be possible to set the Cache-Control headers on the side of Meteor, but am not sure where to look; any pointers?

**Comment by filipenevola** [comment]
@KoenLav I'm investigating a new HCP problem on iOS that we are having but I already change the Cache-Control in my Meteor app for static assets like this

```
WebApp.rawConnectHandlers.use((req, res, next) => {
  if (
    req._parsedUrl.pathname.match(/\.(jpe?g|png|gif|svg|eot|ttf|woff|woff2)$/i)
  ) {
    // one year
    res.setHeader('Cache-Control', `max-age=${OFFLINE_CACHE_EXPIRE}`);
  }
  next();
});
```

**Comment by filipenevola** [comment]
I added this but my HCP still not working on iOS (Meteor 1.8) but it was working until a few weeks ago, then I guess my problem is different

```

  if (req._parsedUrl.pathname.match(/manifest\.json$/i)) {
    // https://github.com/zodern/meteor-up/issues/1086
    // no cache for /__cordova/manifest.json
    res.setHeader(
      'Cache-Control',
      'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0'
    );
  }
```

**Comment by wildhart** [comment]
@filipenevola, I'm pretty sure the `/__cordova` files are not served through the `WebApp` handlers.

I've confirmed this on localhost by using
```js
WebApp.connectHandlers.use((req, res, next) => {
	console.log(req._parsedUrl.pathname);
	next();
});
```
And the lack of console messages confirms that the handler is not being called on the `/__cordova/manifest.json` file.

You can check for yourself: in Chrome navigate direct to your manifest.json file and look at the network tab of Dev Tools.  You will only see `cache-control: public, max-age=31536000`

The extra `Cache-Control` header needs to be added by the server, unless something can be done at a lower level in Meteor by @benjamn...

**Comment by filipenevola** [comment]
@wildhart I checked this but I got the opposite. I was able to see it affecting the requests on Chrome and also logs in the server.

Are you importing the file with this WebApp.connectHandlers code on server side startup?

**Comment by wildhart** [comment]
@filipenevola, yes my test `WebApp.connectHandlers` code is imported into the server bundle.   I can see the console messages when I browse to routes in my app but **not** when I browse to the manifest.js file.  Interesting that you do.

So you've confirmed that your server is now sending the manifest with the correct `no-store, no-cache` header, but iOS *still* isn't updating the HCP, correct?  Have you tried uninstalling the iOS app from the device and reinstalling it?  Once an install of the app has cached an old manifest nothing you change on the server will make it get a fresh copy until you delete the app.

**Comment by wildhart** [comment]
Suggestion for @benjamn: Could it be as simple as adding a cache-busting timestamp to the manifest fetch request, making the app always get a fresh copy no matter what caches/proxies are in the way?

**Comment by filipenevola** [comment]
Yes, I had uninstalled but I don't know why it is not working, we are updating to Meteor 1.8.1 to see if it's working there.

**Comment by brucejo75** [comment]
@wildhart, did you fork the webapp package and give your fix a try?

It is pretty simple to do without rebuilding all of meteor:
- clone meteor
- copy the webapp package into your project
- make modifications to the package
- I usually make sure the package is referred to directly in `.meteor/packages`
- give it a build and your new code will be used.

Then if it works, copy your fixes back to a PR branch on your meteor clone and submit a pull request.

**Comment by filipenevola** [comment]
@wildhart my HCP are working now and the only change I did was the manifest cache one, I think something was still caching last time I tried and today I'm really cache free.

Just to be clear for future people, my only change was this on server side of my Meteor app
```
WebApp.rawConnectHandlers.use((req, res, next) => {
  if (req._parsedUrl.pathname.match(/manifest\.json$/i)) {
    // eslint-disable-next-line no-console
    console.log('requesting manifest.json', req._parsedUrl.pathname);
    // https://github.com/zodern/meteor-up/issues/1086
    // no cache for /__cordova/manifest.json
    res.setHeader(
      'Cache-Control',
      'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0'
    );
  }
  next();
});

```

**Comment by KoenLav** [comment]
@filipenevola this seems like a change that should be integrated into Meteor; could you make a PR?

@benjamn could you confirm?

**Comment by filipenevola** [comment]
@KoenLav sure.

Let's wait for Ben first.

**Comment by wildhart** [comment]
@filipenevola's, I've discovered why your solution wasn't having any effect for me.  You were using `WebApp.rawConnectHandlers` but I was using `WebApp.connectHandlers`.  When I use the `raw` version it works fine (on localhost, haven't tried production).

@brucejo75, I haven't tested my suggestion, and unfortunately at the moment I don't really have time to - to test it properly involves spinning up staging servers and compiling apps in xCode all of which is time consuming.  Might have time at the weekend...  I figured that Ben would know instantly whether it would work or not, or if there are any unintended consequences I'm not aware of.

**Comment by wildhart** [comment]
I've discovered that my errant `cache-control: public, max-age=31536000` header was self-inflicted.  I had included is as part of my CDN configuration, in a separate file I found this:
```js
WebApp.rawConnectHandlers.use(function(req, res, next) {
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader("Access-Control-Allow-Methods", "GET,HEAD,OPTIONS,POST,PUT");
    res.setHeader("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept, Authorization");
    if (req._parsedUrl.pathname.match(/\.(js|font\.css|css|jpg|png|ttf|ttc|otf|eot|woff|woff2|json)$/)) {
        res.setHeader('cache-control', 'public, max-age=31536000');
    } 
    next();
});
```
I must've picked that code up from a guide or forum when I first implemented a CDN years ago.

I've now fixed it to this:
```js
WebApp.rawConnectHandlers.use(function(req, res, next) {
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader("Access-Control-Allow-Methods", "GET,HEAD,OPTIONS,POST,PUT");
    res.setHeader("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept, Authorization");
    if (req._parsedUrl.pathname.match(/\.json$/)) {
        res.setHeader('Cache-Control', 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0');
    } else if (req._parsedUrl.pathname.match(/\.(js|font\.css|css|jpg|png|ttf|ttc|otf|eot|woff|woff2)$/)) {
        res.setHeader('cache-control', 'public, max-age=31536000');
    }
    next();
});
```
So apologies for suggesting that Meteor and/or MUP were inserting this header, it was me all along.  Maybe Meteor/MUP don't need to fix anything, or maybe it wold be wise for them to add appropariate default 'no-cache' headers anyway...

**Comment by derwaldgeist** [comment]
Is this still open for a year now?

**Comment by RealHandy** [comment]
No! It’s fixed in v1.7.0 of the plugin. I think other linked tickets give the whole story.

**Comment by filipenevola** [comment]
Hello, AFAIK this is working. 

I'm going to close this but if someone is still having issues please reply here.
